### DISCUSSION TITLE
NextJS app deploy with python script in api routes error

### DISCUSSION THREAD

**Discussion by marcosaurelioo** [vercel/next.js#46223] [Help] [answered]
Basically, I have a python script which I consume using child_process. Locally it works pretty good, but when it goes to deploy, the request return error 500 (spawn python3 enoent). There's some config missing, I think...

Link to the code that reproduces this issue
Repo: https://github.com/marcosaurelioo/AudioScript

**Comment by icyJoseph** [comment, marked as answer]
Looks like, as of now, python3 is not available in the API runtime.


Two things:

First you do:

```js
  const python = spawn('python3', ['script/speech-recognition.py']);
```

But you should use the `process.cwd()` utility function to get a reference to where your deployment is currently sitting, and from there look for `script/speech-recognition`. More info here, https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd

If that alone doesn't fix it, could you move your scripts into `src`? In this case the problem would be that the file is not being included in the serverless environment because it is outside `src`.

**Reply by marcosaurelioo** [reply]
I tried both, but none of it worked. 
here the [commit](https://github.com/marcosaurelioo/AudioScript/commit/1b469e7e59fc15a93573fc7097f75c60b45d80a2) that I made the changes.

**Reply by icyJoseph** [reply]
Are you sure python is available on the Vercel env used by the API routes? 

Maybe it's easier to have an API folder as sibling to src, and put a Python API handler there: https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/python

**Reply by icyJoseph** [reply]
Since you already have a path to the script, try to fs.fileRead and log that to console, or respond the call with the script contents. If that works then, Python is probably not available on that environment.

**Reply by icyJoseph** [reply]
@marcosaurelioo I was able to read the file on Vercel, so I guess the problem is that Python3 is not available:

![Screenshot 2023-02-22 at 16 55 28](https://user-images.githubusercontent.com/21013447/220680333-c9c2f8c2-2693-424c-8d25-43aa783e3a80.png)

**Reply by marcosaurelioo** [reply]
Yeah, I was getting this error and also presuming the python was not available.
![image](https://user-images.githubusercontent.com/48496372/220680956-2e5d4ebb-6df7-4eb7-a88f-bc2bc74efe3d.png)

There's no way to fix it, right?
I'll do as you said here:
> Are you sure python is available on the Vercel env used by the API routes?
> 
> Maybe it's easier to have an API folder as sibling to src, and put a Python API handler there: https://vercel.com/docs/concepts/functions/serverless-functions/runtimes/python

**Reply by icyJoseph** [reply]
This is now in territory I don't know much about. What's exactly available in the Vercel Node.js environment. I guess the easiest way is to just use a Python API handler. It is what fits your use case best too, but locally it will be weird to develop, feels like there should be other options.

- https://vercel.com/docs/concepts/functions/serverless-functions/supported-languages#advanced-python-usage
- https://vercel.com/docs/concepts/deployments/configure-a-build

I wonder if you could provide a command that install python upon build/server start? Sounds too complicated

**Reply by marcosaurelioo** [reply]
Python is not preinstalled on vercel servers, so I think there's nothing to do. Thx for the help.

**Reply by icyJoseph** [reply]
Aw man, that's a slight let down. I'll update my answer then.
